src/integration/integration.service.ts

```diff
@@ -0,0 +1,352 @@
+  buildValidationsValidateStackableParamsForVoucherify,
+  CommercetoolsService,
+} from '../commercetools/commercetools.service';
+import {
+  getCouponsLimit,
+  VoucherifyService,
+} from '../voucherify/voucherify.service';
+import {
+  deserializeCoupons,
+  filterCouponsByLimit,
+  getCouponsFromCart,
+} from './helperFunctions';
+
+@Injectable()
+export class IntegrationService {
+  constructor(
+    private readonly taxCategoriesService: TaxCategoriesService,
+    private readonly typesService: TypesService,
+    private readonly logger: Logger,
+    @Inject(forwardRef(() => VoucherifyConnectorService))
+    private readonly voucherifyConnectorService: VoucherifyConnectorService,
+    private readonly commerceToolsConnectorService: CommercetoolsConnectorService,
+    private readonly productMapper: ProductMapper,
+    private readonly configService: ConfigService,
+    @Inject(forwardRef(() => CommercetoolsService))
+    private readonly commercetoolsService: CommercetoolsService,
+    private readonly voucherifyService: VoucherifyService,
+  ) {}
+
+  public async validateCouponsAndGetAvailablePromotions(
+    cart: Cart,
+    sessionKey?: string | null,
+  ): Promise<ValidateCouponsResult> {
+    const { id, customerId, anonymousId } = cart;
+    const coupons: Coupon[] = getCouponsFromCart(cart);
+    let uniqCoupons: Coupon[] = uniqBy(coupons, 'code');
+    if (coupons.length !== uniqCoupons.length) {
+      this.logger.debug({
+        msg: 'Duplicates found and deleted',
+      });
+    }
+
+    const { promotions, availablePromotions } =
+      await this.voucherifyService.getPromotions(cart, uniqCoupons);
+
+    if (!uniqCoupons.length) {
+      this.logger.debug({
+        msg: 'No coupons applied, skipping voucherify call',
+      });
+
+      return {
+        availablePromotions,
+      };
+    }
+
+    const deletedCoupons = uniqCoupons.filter(
+      (coupon) => coupon.status === 'DELETED',
+    );
+
+    deletedCoupons
+      .filter((coupon) => coupon.type !== 'promotion_tier')
+      .map((coupon) =>
+        this.voucherifyConnectorService.releaseValidationSession(
+          coupon.code,
+          sessionKey,
+        ),
+      );
+
+    if (deletedCoupons.length === uniqCoupons.length) {
+      this.logger.debug({
+        msg: 'Deleting coupons only, skipping voucherify call',
+      });
+
+      return {
+        availablePromotions,
+      };
+    }
+
+    this.logger.debug({
+      msg: 'Attempt to apply coupons',
+      coupons: uniqCoupons,
+      id,
+      customerId,
+      anonymousId,
+    });
+
+    uniqCoupons = filterCouponsByLimit(
+      uniqCoupons,
+      getCouponsLimit(
+        this.configService.get<number>('COMMERCE_TOOLS_COUPONS_LIMIT'),
+      ),
+    );
+
+    let validatedCoupons: ValidationValidateStackableResponse =
+      await this.voucherifyConnectorService.validateStackableVouchers(
+        buildValidationsValidateStackableParamsForVoucherify(
```

Comment:
Integration service should not operate on types from commercetools. You will rid off `buildValidationsValidateStackableParamsForVoucherify`